### PR TITLE
Add configuration values to tweak the behavior of signature file endpoints

### PR DESCRIPTION
**Pull request by joelverhagen**
The goal here is to allow users of NuGet.Server to set configuration values to mimic nuget.org behavior. This is already the case in `dev` and `master` via the `enableDelisting` configuration value. Three configuration values are added to aid in this goal. Although nuget.org is not built on NuGet.Server, the following settings can be used to make NuGet.Server match nuget.org's signature file behavior:

1. `maxSignatureFiles` - nuget.org will have this set to `1`.
1. `allowSignatureFileDelete` - nuget.org will have this set to `false`.
1. `allowSignatureFileReplace` - nuget.org will have this set to `false`.

### CODE REVIEW

File: samples/NuGet.Server.V2.Samples.OwinHost/Program.cs [hunk 1]
```diff
@@ -24,11 +25,13 @@ static void Main(string[] args)
...
-            settings.Add("enableDelisting", false);                         //default=false
-            settings.Add("enableFrameworkFiltering", false);                //default=false
-            settings.Add("ignoreSymbolsPackages", true);                    //default=false
-            settings.Add("allowOverrideExistingPackageOnPush", true);       //default=true
+            var settings = new Dictionary<string, string>
+            {
+                { "enableDelisting", "false" },
+                { "enableFrameworkFiltering", "false" },
+                { "ignoreSymbolsPackages", "true" },
+                { "allowOverrideExistingPackageOnPush", "true" }
```
Comment (dtivel) [inline samples/NuGet.Server.V2.Samples.OwinHost/Program.cs]: Override => Overwrite

File: src/NuGet.Server/Web.config [hunk 2]
```diff
@@ -36,6 +36,29 @@
...
+    Set allowSignatureFileDelete to true to allow signature files to be deleted by name. If this setting is set to
+    false, HTTP 405 Method Not Allowed is returned when a user tries to delete a signature file. This setting does not
+    effect the case when a package and its signatures is being replaced with "nuget.exe push". 
+    -->
+    <add key="allowSignatureFileDelete" value="true" />
+    
+    <!--
+    Set allowSignatureFileReplace to true to allow signature files to be replaced by name. If this setting is set to
+    false, HTTP 405 Method Not Allowed is returned when a user tries to replace a signature file. This setting does not
+    effect the case when a package and its signatures is being replaced with "nuget.exe push".
```
Comment (dtivel) [inline src/NuGet.Server/Web.config]: effect => affect

File: src/NuGet.Server/Web.config [hunk 3]
```diff
@@ -36,6 +36,29 @@
...
+    <!--
+    Set allowSignatureFileDelete to true to allow signature files to be deleted by name. If this setting is set to
+    false, HTTP 405 Method Not Allowed is returned when a user tries to delete a signature file. This setting does not
+    effect the case when a package and its signatures is being replaced with "nuget.exe push". 
+    -->
+    <add key="allowSignatureFileDelete" value="true" />
+    
+    <!--
+    Set allowSignatureFileReplace to true to allow signature files to be replaced by name. If this setting is set to
+    false, HTTP 405 Method Not Allowed is returned when a user tries to replace a signature file. This setting does not
```
Comment (dtivel) [inline src/NuGet.Server/Web.config]: The spec says 409, but I think 405 is more appropriate.

File: src/NuGet.Server/Web.config [hunk 4]
```diff
@@ -36,6 +36,29 @@
...
+    the value to 0 has the effect of not allowing any signature files on the server. If the value is set to a negative
+    number, there is no limit on the number of signature files. If this setting is changed when there are packages that
+    already have signature files, only future signature file additions will be limited. In other words, this setting is
+    not applied to pre-existing signature files or signature file replace operations.
+    -->
+    <add key="maxSignatureFiles" value="-1" />
+
+    <!--
+    Set allowSignatureFileDelete to true to allow signature files to be deleted by name. If this setting is set to
+    false, HTTP 405 Method Not Allowed is returned when a user tries to delete a signature file. This setting does not
```
Comment (dtivel) [inline src/NuGet.Server/Web.config]: The spec doesn't cover this error case, but I think it should.

File: test/NuGet.Server.Core.Tests/ServerPackageRepositoryTest.cs [hunk 5]
```diff
@@ -1271,6 +1317,88 @@ public class ServerPackageRepositoryTest
...
+                }
+            }
+        }
+
+        [Theory]
+        [InlineData(0, "The NuGet server has been configured to not allow signature files.")]
+        [InlineData(1, "The NuGet server has been configured allow a maximum of 1 signature file.")]
+        [InlineData(2, "The NuGet server has been configured allow a maximum of 2 signature files.")]
+        [InlineData(3, "The NuGet server has been configured allow a maximum of 3 signature files.")]
+        public async Task ServerPackageRepository_AddPackageAndTooManySignatureFiles(int max, string expectedMessage)
```
Comment (dtivel) [inline test/NuGet.Server.Core.Tests/ServerPackageRepositoryTest.cs]: The first 3 tests cover all equivalence categories.  The fourth test seems redundant with the third test.

File: samples/NuGet.Server.V2.Samples.OwinHost/Program.cs (hunk 1 above)
Comment (joelverhagen) [inline samples/NuGet.Server.V2.Samples.OwinHost/Program.cs]: `allowOverrideExistingPackageOnPush` is actually the right configuration key ([source on `master`](https://github.com/NuGet/NuGet.Server/blob/e7807dbdd361fb19a458e353b3e754e8ce61378c/src/NuGet.Server/Web.config#L25)). It's unfortunate but we can't break people.

File: src/NuGet.Server/Web.config (hunk 3 above)
Comment (joelverhagen) [inline src/NuGet.Server/Web.config]: Under the **Replacing an existing signature file** section, it mentions:

> If this operation is not supported by the NuGet server, HTTP 405 Method Not Allowed should be returned.

File: src/NuGet.Server/Web.config (hunk 4 above)
Comment (joelverhagen) [inline src/NuGet.Server/Web.config]: It does in under the **Deleting an existing signature file** section:

> If this operation is not supported by the NuGet server, HTTP 405 Method Not Allowed should be returned.

File: test/NuGet.Server.Core.Tests/ServerPackageRepositoryTest.cs (hunk 5 above)
Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/ServerPackageRepositoryTest.cs]: I am not sure what you mean by "equivalence categories", but I agree with the second point -- redundant test.

File: src/NuGet.Server/Web.config (hunk 2 above)
Comment (joelverhagen) [inline src/NuGet.Server/Web.config]: Fixed.

File: test/NuGet.Server.Core.Tests/ServerPackageRepositoryTest.cs (hunk 5 above)
Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/ServerPackageRepositoryTest.cs]: Removed redundant test.

Comment (dtivel) [inline test/NuGet.Server.Core.Tests/ServerPackageRepositoryTest.cs]: Sorry, "equivalence class."  It's a testing concept you already understand intuitively.

Suppose you're testing a function that accepts integers.  You don't want to test every integer, so you test integers that are interesting.

Equivalence classes might be:  positive integers, negative integers, zero, and extremes.

Actual values might be:  1, -1, 0, int.MaxValue and int.MinValue.

There's no point in also testing and 7 and -19.

Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/ServerPackageRepositoryTest.cs]: Ah, that's what those categories are called 😄. Thanks for the explanation.

**Pull request merged**